examples/darwin-framework-tool/commands/configuration/PrintDeviceCommand.h

```diff
@@ -0,0 +1,443 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+#pragma once
+
+#import <Matter/Matter.h>
+
+#import "../common/DeviceDelegate.h"
+#import "MTRError_Utils.h"
+
+#include <commands/common/Command.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+class PrintDeviceCommand : public CHIPCommandBridge {
+public:
+    PrintDeviceCommand()
+        : CHIPCommandBridge("print", "Prints the endpoints, clusters, commands and attributes that are "
+                                     "available on a device instance.")
+    {
+        AddArgument("node-id", 0, UINT64_MAX, &mNodeId, "The Node ID of the device whose details will be printed.");
+        AddArgument("endpoint-id", 0, UINT16_MAX, &mEndPointId, "If provided, restrict the output to this single endpoint.");
+    }
+
+protected:
+    /////////// CHIPCommandBridge Interface /////////
+    CHIP_ERROR RunCommand() override
+    {
+        auto * controller = CurrentCommissioner();
+        VerifyOrReturnError(nil != controller, CHIP_ERROR_INCORRECT_STATE);
+
+        auto * device = [MTRDevice deviceWithNodeID:@(mNodeId) controller:controller];
+        VerifyOrReturnError(nil != device, CHIP_ERROR_INCORRECT_STATE);
+
+        DeviceWithNodeId(mNodeId);
+
+        if (device.deviceCachePrimed) {
+            return PrintDevice(device);
+        }
+
+        auto onDeviceCachePrimedBlock = ^{
+            PrintDevice(device);
+
+            [device removeDelegate:mDeviceDelegate];
+            mDeviceDelegate = nil;
+        };
+
+        mDeviceDelegate = [[DeviceDelegate alloc] init];
+        mDeviceDelegate.onDeviceCachePrimed = onDeviceCachePrimedBlock;
+        auto deviceDelegateDispatchQueue = dispatch_queue_create("com.chip.devicedelegate", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
+        [device addDelegate:mDeviceDelegate queue:deviceDelegateDispatchQueue];
+
+        return CHIP_NO_ERROR;
+    }
+
+    chip::System::Clock::Timeout GetWaitDuration() const override { return chip::System::Clock::Seconds16(30); }
+
+private:
+    // Error Descriptions
+    NSString * const kErrorDomain = @"Error";
+    NSInteger const kErrorCode = 0;
+    NSString * const kErrorNotAnArray = @"Expected an array, but received a different type.";
+    NSString * const kErrorNotANumber = @"Expected a numeric value, but received a different type.";
+    NSString * const kErrorNotAStructure = @"Expected a structure, but received a different type.";
+
+    CHIP_ERROR PrintDevice(MTRDevice * device)
+    {
+        CHIP_ERROR error = CHIP_NO_ERROR;
+
+        NSError * internalError = nil;
+        AddDeviceToBuffer(device, &internalError);
+        if (nil != internalError) {
+            error = MTRErrorToCHIPErrorCode(internalError);
+            ChipLogError(chipTool, "Error while reading device configuration: %s", internalError.localizedDescription.UTF8String);
+        }
+
+        SetCommandExitStatus(error);
+        return error;
```

Comment:
The documentation for RunCommand says:

```
    // 1) If error is returned, Run() must not call SetCommandExitStatus.
    // 2) If success is returned Run() must either have called
    //    SetCommandExitStatus() or scheduled async work that will do that.
```

This is violating that API contract when calling PrintDevice sync and we get an error, no?